crates/wasi/src/preview2/preview1/mod.rs

```diff
@@ -253,143 +743,599 @@ impl<
+        let fs_filetype = self
+            .get_type(fd)
+            .await
+            .map_err(|e| {
+                e.try_into()
+                    .context("failed to call `get-type`")
+                    .unwrap_or_else(types::Error::trap)
+            })?
+            .try_into()
+            .map_err(types::Error::trap)?;
+        let mut fs_flags = types::Fdflags::empty();
+        let mut fs_rights_base = types::Rights::all();
+        if !flags.contains(wasi::filesystem::DescriptorFlags::READ) {
+            fs_rights_base &= !types::Rights::FD_READ;
+        }
+        if !flags.contains(wasi::filesystem::DescriptorFlags::WRITE) {
+            fs_rights_base &= !types::Rights::FD_WRITE;
+        }
+        if flags.contains(wasi::filesystem::DescriptorFlags::DATA_INTEGRITY_SYNC) {
+            fs_flags |= types::Fdflags::DSYNC;
+        }
+        if flags.contains(wasi::filesystem::DescriptorFlags::REQUESTED_WRITE_SYNC) {
+            fs_flags |= types::Fdflags::RSYNC;
+        }
+        if flags.contains(wasi::filesystem::DescriptorFlags::FILE_INTEGRITY_SYNC) {
+            fs_flags |= types::Fdflags::SYNC;
+        }
+        if append {
+            fs_flags |= types::Fdflags::APPEND;
+        }
+        if !blocking {
+            fs_flags |= types::Fdflags::NONBLOCK;
+        }
+        Ok(types::Fdstat {
+            fs_filetype,
+            fs_flags,
+            fs_rights_base,
+            fs_rights_inheriting: fs_rights_base,
+        })
     }
 
+    /// Adjust the flags associated with a file descriptor.
+    /// NOTE: This is similar to `fcntl(fd, F_SETFL, flags)` in POSIX.
+    #[instrument(skip(self))]
     async fn fd_fdstat_set_flags(
         &mut self,
         fd: types::Fd,
         flags: types::Fdflags,
     ) -> Result<(), types::Error> {
-        todo!()
+        let mut ds = self.descriptors().await?;
+        let File {
+            append, blocking, ..
+        } = ds.get_file_mut(fd)?;
+
+        // Only support changing the NONBLOCK or APPEND flags.
+        if flags.contains(types::Fdflags::DSYNC)
+            || flags.contains(types::Fdflags::SYNC)
+            || flags.contains(types::Fdflags::RSYNC)
+        {
+            return Err(types::Errno::Inval.into());
+        }
+        *append = flags.contains(types::Fdflags::APPEND);
+        *blocking = !flags.contains(types::Fdflags::NONBLOCK);
+        Ok(())
     }
 
+    /// Adjust the rights associated with a file descriptor.
+    /// This can only be used to remove rights, and returns `errno::notcapable` if called in a way that would attempt to add rights
+    #[instrument(skip(self))]
     async fn fd_fdstat_set_rights(
         &mut self,
-        fd: types::Fd,
-        fs_rights_base: types::Rights,
-        fs_rights_inheriting: types::Rights,
+        _fd: types::Fd,
+        _fs_rights_base: types::Rights,
+        _fs_rights_inheriting: types::Rights,
     ) -> Result<(), types::Error> {
-        todo!()
+        unreachable!()
```

Comment:
this is still a `todo!()` - we can leave it to a follow up PR, but `unreachable!()` communicates to me that it is a bug in this code if it is reachable, while this is reachable from any guest program.

Reply:
I took this directly from https://github.com/bytecodealliance/wasmtime/blob/5b93cdb84def5fa7cb3ff823bc80216f19c296c5/crates/wasi-preview1-component-adapter/src/lib.rs#L560-L569 and it was surprising to me as well.
It seems to me that the adapter has to implement rights handling, which the existing one does not do. I'll mark this as `todo!` and file an issue for rights handling follow-up

Reply:
Oh, that is indeed a bug in the component adapter as well. `unreachabable!()` is just how we spell `todo!()` and `panic!()` and etc over there since we cant use std's macros.

In the legacy implementation, set_rights is a no-op, it only validates the fd. That is the correct behavior for both of these spots. Would you mind making this fix to the component adapter as well?

Reply:
Let's take care of it as part of https://github.com/bytecodealliance/wasmtime/issues/6446 ?